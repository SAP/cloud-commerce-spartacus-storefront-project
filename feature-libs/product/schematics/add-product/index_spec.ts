import {
  SchematicTestRunner,
  UnitTestTree,
} from '@angular-devkit/schematics/testing';
import {
  LibraryOptions as SpartacusProductOptions,
  SpartacusOptions,
} from '@spartacus/schematics';
import * as path from 'path';
import {
  SPARTACUS_BULK_PRICING_ROOT,
  CLI_BULK_PRICING_FEATURE,
  SPARTACUS_VARIANTS_ROOT,
  CLI_VARIANTS_FEATURE,
} from '../constants';

const collectionPath = path.join(__dirname, '../collection.json');
const appModulePath = 'src/app/app.module.ts';

describe('Spartacus Product schematics: ng-add', () => {
  const schematicRunner = new SchematicTestRunner('schematics', collectionPath);

  let appTree: UnitTestTree;

  const workspaceOptions: any = {
    name: 'workspace',
    version: '0.5.0',
  };

  const appOptions: any = {
    name: 'schematics-test',
    inlineStyle: false,
    inlineTemplate: false,
    routing: false,
    style: 'scss',
    skipTests: false,
    projectRoot: '',
  };

  const defaultOptions: SpartacusProductOptions = {
    project: 'schematics-test',
    lazy: true,
    features: [CLI_BULK_PRICING_FEATURE, CLI_VARIANTS_FEATURE],
  };

  const spartacusDefaultOptions: SpartacusOptions = {
    project: 'schematics-test',
  };

  beforeEach(async () => {
    schematicRunner.registerCollection(
      '@spartacus/schematics',
      '../../projects/schematics/src/collection.json'
    );
    schematicRunner.registerCollection(
      '@spartacus/storefinder',
      '../../feature-libs/storefinder/schematics/collection.json'
    );

    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'workspace',
        workspaceOptions
      )
      .toPromise();
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'application',
        appOptions,
        appTree
      )
      .toPromise();
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@spartacus/schematics',
        'ng-add',
        { ...spartacusDefaultOptions, name: 'schematics-test' },
        appTree
      )
      .toPromise();
  });

  describe('when no features are selected', () => {
    beforeEach(async () => {
      appTree = await schematicRunner
        .runSchematicAsync(
          'ng-add',
          { ...defaultOptions, features: [] },
          appTree
        )
        .toPromise();
    });

    it('should not install bulkPricing feature', () => {
      const appModule = appTree.readContent(appModulePath);
      expect(appModule).not.toContain(SPARTACUS_BULK_PRICING_ROOT);
    });

    it('should not install variants feature', () => {
      const appModule = appTree.readContent(appModulePath);
      expect(appModule).not.toContain(SPARTACUS_VARIANTS_ROOT);
    });
  });

  describe('when other Spartacus features are already installed', () => {
    beforeEach(async () => {
      appTree = await schematicRunner
        .runExternalSchematicAsync(
          '@spartacus/storefinder',
          'ng-add',
          { ...spartacusDefaultOptions, name: 'schematics-test' },
          appTree
        )
        .toPromise();
      appTree = await schematicRunner
        .runSchematicAsync('ng-add', defaultOptions, appTree)
        .toPromise();
    });

    it('should just append the product features without duplicating the featureModules config', () => {
      const appModule = appTree.readContent(appModulePath);
      expect(appModule.match(/featureModules:/g)?.length).toEqual(1);
      expect(appModule).toContain(`bulkPricing: {`);
      expect(appModule).toContain(`variants: {`);
    });
  });

  describe('styling', () => {
    beforeEach(async () => {
      appTree = await schematicRunner
        .runSchematicAsync('ng-add', defaultOptions, appTree)
        .toPromise();
    });

    it('should add style import to /src/styles/spartacus/product.scss', async () => {
      const content = appTree.readContent('/src/styles/spartacus/product.scss');
      expect(content).toEqual(`@import "@spartacus/product";`);
    });

    it('should add update angular.json with spartacus/product.scss', async () => {
      const content = appTree.readContent('/angular.json');
      const angularJson = JSON.parse(content);
      const buildStyles: string[] =
        angularJson.projects['schematics-test'].architect.build.options.styles;
      expect(buildStyles).toEqual([
        'src/styles.scss',
        'src/styles/spartacus/product.scss',
      ]);

      const testStyles: string[] =
        angularJson.projects['schematics-test'].architect.test.options.styles;
      expect(testStyles).toEqual([
        'src/styles.scss',
        'src/styles/spartacus/product.scss',
      ]);
    });
  });
});
