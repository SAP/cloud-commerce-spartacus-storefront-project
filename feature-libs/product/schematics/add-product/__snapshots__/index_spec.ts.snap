// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spartacus Product schematics: ng-add BulkPricing feature eager loading should import appropriate modules 1`] = `
"import { NgModule } from '@angular/core';
import { I18nConfig, provideConfig } from \\"@spartacus/core\\";
import { BulkPricingModule } from \\"@spartacus/product/bulk-pricing\\";
import { bulkPricingTranslationChunksConfig, bulkPricingTranslations } from \\"@spartacus/product/bulk-pricing/assets\\";
import { BulkPricingRootModule } from \\"@spartacus/product/bulk-pricing/root\\";

@NgModule({
  declarations: [],
  imports: [
    BulkPricingRootModule,
    BulkPricingModule
  ],
  providers: [provideConfig(<I18nConfig>{
    i18n: {
      resources: bulkPricingTranslations,
      chunks: bulkPricingTranslationChunksConfig,
    },
  })]
})
export class ProductBulkPricingFeatureModule { }
"
`;

exports[`Spartacus Product schematics: ng-add BulkPricing feature general setup should add the feature using the lazy loading syntax 1`] = `
"import { NgModule } from '@angular/core';
import { CmsConfig, I18nConfig, provideConfig } from \\"@spartacus/core\\";
import { bulkPricingTranslationChunksConfig, bulkPricingTranslations } from \\"@spartacus/product/bulk-pricing/assets\\";
import { BulkPricingRootModule, PRODUCT_BULK_PRICING_FEATURE } from \\"@spartacus/product/bulk-pricing/root\\";

@NgModule({
  declarations: [],
  imports: [
    BulkPricingRootModule
  ],
  providers: [provideConfig(<CmsConfig>{
    featureModules: {
      [PRODUCT_BULK_PRICING_FEATURE]: {
        module: () =>
          import('@spartacus/product/bulk-pricing').then((m) => m.BulkPricingModule),
      },
    }
  }),
  provideConfig(<I18nConfig>{
    i18n: {
      resources: bulkPricingTranslations,
      chunks: bulkPricingTranslationChunksConfig,
    },
  })
  ]
})
export class ProductBulkPricingFeatureModule { }
"
`;

exports[`Spartacus Product schematics: ng-add BulkPricing feature general setup styling should create a proper scss file 1`] = `"@import \\"@spartacus/product\\";"`;

exports[`Spartacus Product schematics: ng-add BulkPricing feature general setup styling should update angular.json 1`] = `
"{
  \\"$schema\\": \\"./node_modules/@angular/cli/lib/config/schema.json\\",
  \\"version\\": 1,
  \\"newProjectRoot\\": \\"\\",
  \\"projects\\": {
    \\"schematics-test\\": {
      \\"projectType\\": \\"application\\",
      \\"schematics\\": {
        \\"@schematics/angular:component\\": {
          \\"style\\": \\"scss\\"
        }
      },
      \\"root\\": \\"\\",
      \\"sourceRoot\\": \\"src\\",
      \\"prefix\\": \\"app\\",
      \\"architect\\": {
        \\"build\\": {
          \\"builder\\": \\"@angular-devkit/build-angular:browser\\",
          \\"options\\": {
            \\"outputPath\\": \\"dist/schematics-test\\",
            \\"index\\": \\"src/index.html\\",
            \\"main\\": \\"src/main.ts\\",
            \\"polyfills\\": \\"src/polyfills.ts\\",
            \\"tsConfig\\": \\"tsconfig.app.json\\",
            \\"aot\\": true,
            \\"assets\\": [
              \\"src/favicon.ico\\",
              \\"src/assets\\"
            ],
            \\"styles\\": [
              \\"src/styles.scss\\",
              \\"src/styles/spartacus/product.scss\\"
            ],
            \\"scripts\\": []
          },
          \\"configurations\\": {
            \\"production\\": {
              \\"fileReplacements\\": [
                {
                  \\"replace\\": \\"src/environments/environment.ts\\",
                  \\"with\\": \\"src/environments/environment.prod.ts\\"
                }
              ],
              \\"optimization\\": true,
              \\"outputHashing\\": \\"all\\",
              \\"sourceMap\\": false,
              \\"extractCss\\": true,
              \\"namedChunks\\": false,
              \\"extractLicenses\\": true,
              \\"vendorChunk\\": false,
              \\"buildOptimizer\\": true,
              \\"budgets\\": [
                {
                  \\"type\\": \\"initial\\",
                  \\"maximumWarning\\": \\"2mb\\",
                  \\"maximumError\\": \\"5mb\\"
                },
                {
                  \\"type\\": \\"anyComponentStyle\\",
                  \\"maximumWarning\\": \\"6kb\\",
                  \\"maximumError\\": \\"10kb\\"
                }
              ]
            }
          }
        },
        \\"serve\\": {
          \\"builder\\": \\"@angular-devkit/build-angular:dev-server\\",
          \\"options\\": {
            \\"browserTarget\\": \\"schematics-test:build\\"
          },
          \\"configurations\\": {
            \\"production\\": {
              \\"browserTarget\\": \\"schematics-test:build:production\\"
            }
          }
        },
        \\"extract-i18n\\": {
          \\"builder\\": \\"@angular-devkit/build-angular:extract-i18n\\",
          \\"options\\": {
            \\"browserTarget\\": \\"schematics-test:build\\"
          }
        },
        \\"test\\": {
          \\"builder\\": \\"@angular-devkit/build-angular:karma\\",
          \\"options\\": {
            \\"main\\": \\"src/test.ts\\",
            \\"polyfills\\": \\"src/polyfills.ts\\",
            \\"tsConfig\\": \\"tsconfig.spec.json\\",
            \\"karmaConfig\\": \\"karma.conf.js\\",
            \\"assets\\": [
              \\"src/favicon.ico\\",
              \\"src/assets\\"
            ],
            \\"styles\\": [
              \\"src/styles.scss\\",
              \\"src/styles/spartacus/product.scss\\"
            ],
            \\"scripts\\": []
          }
        },
        \\"lint\\": {
          \\"builder\\": \\"@angular-devkit/build-angular:tslint\\",
          \\"options\\": {
            \\"tsConfig\\": [
              \\"tsconfig.app.json\\",
              \\"tsconfig.spec.json\\",
              \\"e2e/tsconfig.json\\"
            ],
            \\"exclude\\": [
              \\"**/node_modules/**\\"
            ]
          }
        },
        \\"e2e\\": {
          \\"builder\\": \\"@angular-devkit/build-angular:protractor\\",
          \\"options\\": {
            \\"protractorConfig\\": \\"e2e/protractor.conf.js\\",
            \\"devServerTarget\\": \\"schematics-test:serve\\"
          },
          \\"configurations\\": {
            \\"production\\": {
              \\"devServerTarget\\": \\"schematics-test:serve:production\\"
            }
          }
        }
      }
    }
  },
  \\"defaultProject\\": \\"schematics-test\\"
}"
`;

exports[`Spartacus Product schematics: ng-add Variants feature eager loading should import appropriate modules 1`] = `
"import { NgModule } from '@angular/core';
import { I18nConfig, provideConfig } from \\"@spartacus/core\\";
import { ProductVariantsModule } from \\"@spartacus/product/variants\\";
import { productVariantsTranslationChunksConfig, productVariantsTranslations } from \\"@spartacus/product/variants/assets\\";
import { ProductVariantsRootModule } from \\"@spartacus/product/variants/root\\";

@NgModule({
  declarations: [],
  imports: [
    ProductVariantsRootModule,
    ProductVariantsModule
  ],
  providers: [provideConfig(<I18nConfig>{
    i18n: {
      resources: productVariantsTranslations,
      chunks: productVariantsTranslationChunksConfig,
    },
  })]
})
export class ProductVariantsFeatureModule { }
"
`;

exports[`Spartacus Product schematics: ng-add Variants feature general setup should add the feature using the lazy loading syntax 1`] = `
"import { NgModule } from '@angular/core';
import { CmsConfig, I18nConfig, provideConfig } from \\"@spartacus/core\\";
import { productVariantsTranslationChunksConfig, productVariantsTranslations } from \\"@spartacus/product/variants/assets\\";
import { ProductVariantsRootModule, PRODUCT_VARIANTS_FEATURE } from \\"@spartacus/product/variants/root\\";

@NgModule({
  declarations: [],
  imports: [
    ProductVariantsRootModule
  ],
  providers: [provideConfig(<CmsConfig>{
    featureModules: {
      [PRODUCT_VARIANTS_FEATURE]: {
        module: () =>
          import('@spartacus/product/variants').then((m) => m.ProductVariantsModule),
      },
    }
  }),
  provideConfig(<I18nConfig>{
    i18n: {
      resources: productVariantsTranslations,
      chunks: productVariantsTranslationChunksConfig,
    },
  })
  ]
})
export class ProductVariantsFeatureModule { }
"
`;

exports[`Spartacus Product schematics: ng-add Variants feature general setup styling should create a proper scss file 1`] = `"@import \\"@spartacus/product\\";"`;

exports[`Spartacus Product schematics: ng-add Variants feature general setup styling should update angular.json 1`] = `
"{
  \\"$schema\\": \\"./node_modules/@angular/cli/lib/config/schema.json\\",
  \\"version\\": 1,
  \\"newProjectRoot\\": \\"\\",
  \\"projects\\": {
    \\"schematics-test\\": {
      \\"projectType\\": \\"application\\",
      \\"schematics\\": {
        \\"@schematics/angular:component\\": {
          \\"style\\": \\"scss\\"
        }
      },
      \\"root\\": \\"\\",
      \\"sourceRoot\\": \\"src\\",
      \\"prefix\\": \\"app\\",
      \\"architect\\": {
        \\"build\\": {
          \\"builder\\": \\"@angular-devkit/build-angular:browser\\",
          \\"options\\": {
            \\"outputPath\\": \\"dist/schematics-test\\",
            \\"index\\": \\"src/index.html\\",
            \\"main\\": \\"src/main.ts\\",
            \\"polyfills\\": \\"src/polyfills.ts\\",
            \\"tsConfig\\": \\"tsconfig.app.json\\",
            \\"aot\\": true,
            \\"assets\\": [
              \\"src/favicon.ico\\",
              \\"src/assets\\"
            ],
            \\"styles\\": [
              \\"src/styles.scss\\",
              \\"src/styles/spartacus/product.scss\\"
            ],
            \\"scripts\\": []
          },
          \\"configurations\\": {
            \\"production\\": {
              \\"fileReplacements\\": [
                {
                  \\"replace\\": \\"src/environments/environment.ts\\",
                  \\"with\\": \\"src/environments/environment.prod.ts\\"
                }
              ],
              \\"optimization\\": true,
              \\"outputHashing\\": \\"all\\",
              \\"sourceMap\\": false,
              \\"extractCss\\": true,
              \\"namedChunks\\": false,
              \\"extractLicenses\\": true,
              \\"vendorChunk\\": false,
              \\"buildOptimizer\\": true,
              \\"budgets\\": [
                {
                  \\"type\\": \\"initial\\",
                  \\"maximumWarning\\": \\"2mb\\",
                  \\"maximumError\\": \\"5mb\\"
                },
                {
                  \\"type\\": \\"anyComponentStyle\\",
                  \\"maximumWarning\\": \\"6kb\\",
                  \\"maximumError\\": \\"10kb\\"
                }
              ]
            }
          }
        },
        \\"serve\\": {
          \\"builder\\": \\"@angular-devkit/build-angular:dev-server\\",
          \\"options\\": {
            \\"browserTarget\\": \\"schematics-test:build\\"
          },
          \\"configurations\\": {
            \\"production\\": {
              \\"browserTarget\\": \\"schematics-test:build:production\\"
            }
          }
        },
        \\"extract-i18n\\": {
          \\"builder\\": \\"@angular-devkit/build-angular:extract-i18n\\",
          \\"options\\": {
            \\"browserTarget\\": \\"schematics-test:build\\"
          }
        },
        \\"test\\": {
          \\"builder\\": \\"@angular-devkit/build-angular:karma\\",
          \\"options\\": {
            \\"main\\": \\"src/test.ts\\",
            \\"polyfills\\": \\"src/polyfills.ts\\",
            \\"tsConfig\\": \\"tsconfig.spec.json\\",
            \\"karmaConfig\\": \\"karma.conf.js\\",
            \\"assets\\": [
              \\"src/favicon.ico\\",
              \\"src/assets\\"
            ],
            \\"styles\\": [
              \\"src/styles.scss\\",
              \\"src/styles/spartacus/product.scss\\"
            ],
            \\"scripts\\": []
          }
        },
        \\"lint\\": {
          \\"builder\\": \\"@angular-devkit/build-angular:tslint\\",
          \\"options\\": {
            \\"tsConfig\\": [
              \\"tsconfig.app.json\\",
              \\"tsconfig.spec.json\\",
              \\"e2e/tsconfig.json\\"
            ],
            \\"exclude\\": [
              \\"**/node_modules/**\\"
            ]
          }
        },
        \\"e2e\\": {
          \\"builder\\": \\"@angular-devkit/build-angular:protractor\\",
          \\"options\\": {
            \\"protractorConfig\\": \\"e2e/protractor.conf.js\\",
            \\"devServerTarget\\": \\"schematics-test:serve\\"
          },
          \\"configurations\\": {
            \\"production\\": {
              \\"devServerTarget\\": \\"schematics-test:serve:production\\"
            }
          }
        }
      }
    }
  },
  \\"defaultProject\\": \\"schematics-test\\"
}"
`;
